### DISCUSSION TITLE
Running the frontend and api in the same project blocks when the frontend calls an api endpoint.

### DISCUSSION THREAD

**Discussion by clande** [zauberzeug/nicegui#943] [Q&A] [answered]
### Question

I'm new to nicegui and fastapi so I expect this is just a misunderstanding of the threading model.

I started with the [nicegui fastapi example](https://github.com/zauberzeug/nicegui/tree/main/examples/fastapi) with the hope that I could develop the frontend and api in the same project. However when my frontend tries to call the local api endpoint it hangs until it times out. By putting some print statements in I can see that the frontend makes the request but then the api doesn't actually process it until the frontend's request has timed out. So it looks like the first request (constructing the UI) blocks the same thread that the API is trying to use.

I could run the frontend and api as distinct applications but it'd be easier to just run them together if possible.

How do I get this to work? 

[This code](https://github.com/clande/demo) demonstrates the issue.

TIA,
Chris

**Comment by rodja** [comment, marked as answer]
Thanks for reaching out. There are multiple issues and misunderstandings in play. Let's untangle it:

- Your code is blocking because you are using the blocking `requests` library. NiceGUI (and the underlying FastAPI) are async frameworks. That means, no io-bound or cpu-bound tasks should be directly executed on the main thread but rather be "awaited". See https://fastapi.tiangolo.com/async/ for more details. I recommend [httpx as a lib which supports async/await](https://www.python-httpx.org/async/).
- The FastAPI example is overkill if you want to provide RESTful APIs. It's more for cases where you already have an FastAPI backend and want to incorporate some UI elements. Simply use `from nicegui import app` to get the automatically created FastAPI app with which you can define new endpoints.
- Last: You do not need the API implementation at all in your case. NiceGUI follows a backend-first strategy. That means you can directly interact with the UI elements from Python. The UI will be updated automatically. The possibility to define API endpoints is only required to provide data endpoints a native mobile App or other services which need JSON instead of a graphical UI.

**Reply by clande** [reply]
Thanks for your response. That clears several things up for me.
